src/algebra/group/unique_prods.lean

```diff
@@ -0,0 +1,223 @@
+  end⟩
+
+/--  `unique_mul` is preserved by inverse images under injective, multiplicative maps. -/
+@[to_additive "`unique_add` is preserved by inverse images under injective, additive maps."]
+lemma mul_hom_preimage (f : G →ₙ* H) (hf : function.injective f) (a0 b0 : G) {A B : finset H}
+  (u : unique_mul A B (f a0) (f b0)) :
+  unique_mul (A.preimage f (set.inj_on_of_injective hf _))
+    (B.preimage f (set.inj_on_of_injective hf _)) a0 b0 :=
+begin
+  intros a b ha hb ab,
+  rw [← hf.eq_iff, ← hf.eq_iff],
+  rw [← hf.eq_iff, map_mul, map_mul] at ab,
+  exact u (finset.mem_preimage.mp ha) (finset.mem_preimage.mp hb) ab,
+end
+
+/--  `unique_mul` is preserved under multiplicative maps that are injective.
+
+See `unique_mul.mul_hom_map_iff` for a version with swapped bundling. -/
+@[to_additive "`unique_add` is preserved under additive maps that are injective.
+
+See `unique_add.add_hom_map_iff` for a version with swapped bundling."]
+lemma mul_hom_image_iff [decidable_eq H] (f : G →ₙ* H) (hf : function.injective f) :
+  unique_mul (A.image f) (B.image f) (f a0) (f b0) ↔ unique_mul A B a0 b0 :=
+begin
+  refine ⟨λ h, _, λ h, _⟩,
+  { intros a b ha hb ab,
+    rw [← hf.eq_iff, ← hf.eq_iff],
+    rw [← hf.eq_iff, map_mul, map_mul] at ab,
+    exact h (finset.mem_image.mpr ⟨_, ha, rfl⟩) (finset.mem_image.mpr ⟨_, hb, rfl⟩) ab},
+  { intros a b aA bB ab,
+    obtain ⟨a, ha, rfl⟩ : ∃ a' ∈ A, f a' = a := finset.mem_image.mp aA,
+    obtain ⟨b, hb, rfl⟩ : ∃ b' ∈ B, f b' = b := finset.mem_image.mp bB,
+    rw [hf.eq_iff, hf.eq_iff],
+    rw [← map_mul, ← map_mul, hf.eq_iff] at ab,
+    exact h ha hb ab },
+end
+
+/--  `unique_mul` is preserved under embeddings that are multiplicative.
+
+See `unique_mul.mul_hom_image_iff` for a version with swapped bundling. -/
+@[to_additive "`unique_add` is preserved under embeddings that are additive.
+
+See `unique_add.add_hom_image_iff` for a version with swapped bundling."]
+lemma mul_hom_map_iff (f : G ↪ H) (mul : ∀ x y, f (x * y) = f x * f y) :
+  unique_mul (A.map f) (B.map f) (f a0) (f b0) ↔ unique_mul A B a0 b0 :=
+begin
+  classical,
+  convert mul_hom_image_iff ⟨f, mul⟩ f.2;
+  { ext,
+    simp only [finset.mem_map, mul_hom.coe_mk, finset.mem_image] },
+end
+
+end unique_mul
+
+/--  Let `G` be a Type with addition.  `unique_sums G` asserts that any two non-empty
+finite subsets of `A` have the `unique_add` property, with respect to some element of their
+sum `A + B`. -/
+class unique_sums (G) [has_add G] : Prop :=
+( adds : ∀ {A B : finset G} (hA : A.nonempty) (hB : B.nonempty),
+  ∃ (a0 ∈ A) (b0 ∈ B), unique_add A B a0 b0)
+
+/--  Let `G` be a Type with multiplication.  `unique_prods G` asserts that any two non-empty
+finite subsets of `G` have the `unique_mul` property, with respect to some element of their
+product `A * B`. -/
+class unique_prods (G) [has_mul G] : Prop :=
+( muls : ∀ {A B : finset G} (hA : A.nonempty) (hB : B.nonempty),
+  ∃ (a0 ∈ A) (b0 ∈ B), unique_mul A B a0 b0 )
```

Comment:
I was using the name to suggest that the `unique_sums/prods` property holds for all pairs of non-empty sets.  I had thought of calling this `univ`, but I worried that it would create a wrong expectation.

If you have a suggestion for a better name, I'd be happy to follow it!